### PR TITLE
s3 sources: Add testdrive tests and testdrive support

### PR DESCRIPTION
**Pull request by quodlibetor**
The three commits in this PR work together to make it more ergonomic to run testdrive tests
against AWS resources, and to add tests for future AWS resources without hurting the overall ability to have correctness for tests.

Part of #4914 
Tests were split off from #5202 to make them easier to review.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5331)

### CODE REVIEW

File: test/testdrive/mzcompose.yml [hunk 1]
```diff
@@ -8,6 +8,58 @@
...
+
+mzworkflows:
+  # Run tests, requires aws credentials in the environment
+  ci:
+    env:
+      AWS_REGION: us-east-2
+      AWS_ENDPOINT: ''
+      # The first TD_TEST parameter is automatically appended to
+      # test/testdrive, but the second is just a plain string
+      TD_TEST: ${TD_TEST:-*.td test/testdrive/esoteric/*.td}
```
Comment (quodlibetor) [inline test/testdrive/mzcompose.yml:20]: This should be changed, I think either by making testdrive itself handle `TD_TEST` or by adding a `command_target` key to mzcompose's `workflow`

**Review by JLDLaughlin — APPROVED** [review]

File: doc/developer/guide-testing.md [hunk 2]
```diff
@@ -257,11 +259,65 @@ provide most of the coverage, but it's worth adding some (*very*) basic
 testdrive tests (e.g., `> SELECT DATE '1999-01-01'`) to ensure that our pgwire
 implementation is properly serializing dates.
 
-Like the [Unit Tests](#unitintegration-tests), many testdrive tests require
-Zookeeper, Kafka, and the Confluent Schema Registry.
+#### Usage via mzcompose/Docker
```
Comment (benesch) [inline doc/developer/guide-testing.md:262]: Let's just say "Usage via mzcompose" and let the fact that mzcompose is Docker nonsense be documented over in the mzcompose documentation.

File: doc/developer/guide-testing.md [hunk 3]
```diff
@@ -257,11 +259,65 @@ provide most of the coverage, but it's worth adding some (*very*) basic
...
+
+The easiest way to run testdrive tests is via mzcompose.
+
+Testdrive comes with a Docker Compose environment that can set all
+of this up for you, including re-building and running the testdrive
+and materialized binaries if you have changed them locally. From the
+`test/testdrive` directory:
```
Comment (benesch) [inline doc/developer/guide-testing.md:269]: The first sentence in this paragraph no longer makes sense given the reordering, since "all of this" hasn't been defined yet. Can we rephrase to something like this?

> The mzcompose configuration will automatically set up all of testdrive's dependencies, including Zookeeper, Kafka, the Confluent Schema Registry, and mock versions of AWS S3 and Kinesis.

File: doc/developer/guide-testing.md [hunk 4]
```diff
@@ -257,11 +259,65 @@ provide most of the coverage, but it's worth adding some (*very*) basic
...
 
-For testdrive tests that interacts with Amazon Kinesis, you will need to be
-running [LocalStack](https://github.com/localstack/localstack). To install and
+  ```console
+  $ aws-vault exec scratch -- env TD_TEST=**/*.td ./mzcompose run local-aws
+  ```
+
+[aws-creds-docs]: https://handbook.dev.i.mtrlz.dev/setup/#configure-aws-vault
+
+#### Usage without docker
```
Comment (benesch) [inline doc/developer/guide-testing.md:314]: "Usage without mzcompose"

File: doc/developer/guide-testing.md [hunk 5]
```diff
@@ -257,11 +259,65 @@ provide most of the coverage, but it's worth adding some (*very*) basic
...
+  $ TD_TEST=esoteric/*.td ./mzcompose run local-aws
+  ```
+
+* `ci`: Expect actual AWS credentials to be available (q.v. [our documentation][aws-creds-docs]),
+  either in the process environment or via AWS EC2 Profiles:
 
-For testdrive tests that interacts with Amazon Kinesis, you will need to be
-running [LocalStack](https://github.com/localstack/localstack). To install and
+  ```console
+  $ aws-vault exec scratch -- env TD_TEST=**/*.td ./mzcompose run local-aws
```
Comment (benesch) [inline doc/developer/guide-testing.md:309]: These are great, thanks!

File: src/testdrive/src/action.rs [hunk 6]
```diff
@@ -75,15 +78,17 @@ pub struct State {
     aws_credentials: AwsCredentials,
     kinesis_client: KinesisClient,
     kinesis_stream_names: Vec<String>,
+    s3_client: S3Client,
+    s3_buckets_created: BTreeSet<String>,
 }
 
 impl State {
-    pub async fn reset_materialized(&mut self) -> Result<(), Error> {
+    pub(crate) async fn reset_materialized(&mut self) -> Result<(), Error> {
```
Comment (benesch) [inline src/testdrive/src/action.rs:86]: I'd like to avoid changing things to `pub(crate)` arbitrarily. Testdrive is careful to only use `pub(crate)` where doing so actually restricts the visibility of an item. I think it's fine if other crates want to set different standards on that front, but I find confusing, and doubly so when there's no consistent standard within one crate.

File: src/testdrive/src/action.rs [hunk 7]
```diff
@@ -96,33 +101,116 @@ impl State {
...
-                "Deleting Kinesis streams {}",
-                self.kinesis_stream_names.join(", ")
-            );
-            for stream_name in &self.kinesis_stream_names {
-                self.kinesis_client
-                    .delete_stream(DeleteStreamInput {
-                        enforce_consumer_deletion: Some(true),
-                        stream_name: stream_name.clone(),
+    /// Delete the Kinesis streams created for this run of testdrive.
+    pub(crate) async fn reset_kinesis(&mut self) -> Result<(), Error> {
```
Comment (benesch) [inline src/testdrive/src/action.rs:109]: Ditto here.

File: src/testdrive/src/action.rs [hunk 8]
```diff
@@ -471,30 +445,18 @@ pub async fn create_state(
...
-            Vec::new(),
-        )
-    };
+    let aws_info = aws::ConnectInfo::new(
+        config.aws_region.clone(),
+        Some(config.aws_credentials.aws_access_key_id().to_owned()),
+        Some(config.aws_credentials.aws_secret_access_key().to_owned()),
+        config.aws_credentials.token().clone(),
+    )
+    .unwrap();
```
Comment (benesch) [inline src/testdrive/src/action.rs:550]: `.unwrap()`?

File: src/testdrive/src/error.rs [hunk 9]
```diff
@@ -87,12 +107,17 @@ impl Error {
...
                 write_error_heading(&mut stderr, &color_spec)?;
                 write!(&mut stderr, "{}", ctx)?;
-                if let Some(cause) = cause {
-                    writeln!(&mut stderr, ": {}", cause)?;
+                for (i, cause) in causes.iter().enumerate() {
+                    if i == 0 {
+                        write!(&mut stderr, ": {}", cause)?;
+                    } else {
+                        write!(&mut stderr, ", {}", cause)?;
+                    }
```
Comment (benesch) [inline src/testdrive/src/error.rs:119]: Does this display well? I feel like it will output stuff like this that is very hard to parse when errors have nested causes:

```
error: resetting s3 state: failed to delete bucket: http connection broken, failed to delete bucket: network error: peer dropped connection, failed to delete kinesis stream: unknown stream
```

File: src/testdrive/src/error.rs [hunk 10]
```diff
@@ -178,10 +203,16 @@ impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
             Error::Input { err, .. } => write!(f, "{}", err.msg),
-            Error::General { ctx, cause, .. } => {
+            Error::General { ctx, causes, .. } => {
                 write!(f, "{}", ctx)?;
-                if let Some(cause) = cause {
-                    write!(f, ": {}", cause)?;
+                write!(f, "{}", ctx)?;
```
Comment (benesch) [inline src/testdrive/src/error.rs:208]: This line looks like a duplicate of the first?

File: src/testdrive/src/error.rs [hunk 11]
```diff
@@ -47,7 +48,7 @@ pub enum Error {
     },
     General {
         ctx: String,
-        cause: Option<Box<dyn StdError + Send + Sync>>,
+        causes: Vec<DynError>,
```
Comment (benesch) [inline src/testdrive/src/error.rs:51]: I've been kind of hoping to switch testdrive over to anyhow, which this inhibits a bit. I'm not sure that there's a better option, but wanted to get your gears turning.

File: src/testdrive/src/action.rs [hunk 12]
```diff
@@ -120,6 +125,93 @@ impl State {
...
+                        if response.next_continuation_token.is_none() {
+                            break;
+                        }
+                        continuation_token = response.next_continuation_token;
+                    }
+                    Err(e) => {
+                        errors.push(
+                            Error::with_cause(format!("Listing objects for bucket {}", bucket), e)
+                                .into(),
+                        );
```
Comment (benesch) [inline src/testdrive/src/action.rs:187]: This doesn't look quite right to me. If you encounter an error, seems like you'll get stuck in the loop forever.

File: src/testdrive/src/action.rs [hunk 13]
```diff
@@ -120,6 +125,93 @@ impl State {
...
+                }
+            }
+
+            let res = self
+                .s3_client
+                .delete_bucket(DeleteBucketRequest {
+                    bucket,
+                    expected_bucket_owner: None,
+                })
+                .await;
```
Comment (benesch) [inline src/testdrive/src/action.rs:198]: I think it would be substantially clearer if you split everything in this loop body out into its own method like `reset_s3_bucket`. Then you'd be able to just return errors normally from there.

File: src/testdrive/src/action.rs [hunk 14]
```diff
@@ -120,6 +125,93 @@ impl State {
         }
         Ok(())
     }
+
+    /// Delete S3 buckets that were created in this run
+    pub(crate) async fn reset_s3(&mut self) -> Result<(), Error> {
+        if self.s3_buckets_created.is_empty() {
+            return Ok(());
+        }
```
Comment (benesch) [inline src/testdrive/src/action.rs:133]: This check seems redundant given that the rest of the function already won't do anything if `s3_buckets_created` is empty.

File: src/testdrive/src/action.rs [hunk 15]
```diff
@@ -120,6 +125,93 @@ impl State {
...
+
+        let buckets: Vec<String> = self.s3_buckets_created.iter().cloned().collect();
```
Comment (benesch) [inline src/testdrive/src/action.rs:135]: Confused about this. You can just iterate over `&self.s3_buckets_created`.

File: src/testdrive/src/action.rs [hunk 16]
```diff
@@ -120,6 +125,93 @@ impl State {
...
+                errors.push(e.into());
+            }
+        }
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::General {
+                ctx: format!("deleting S3 buckets: {} errors", errors.len()),
+                causes: errors.into_iter().map(Into::into).collect(),
```
Comment (benesch) [inline src/testdrive/src/action.rs:210]: You already called `into` when you pushed into the vec, so you can just do `causes: errors` here.

File: src/testdrive/src/action.rs [hunk 17]
```diff
@@ -458,6 +554,11 @@ pub async fn create_state(
         &[format!("region: {}", aws_info.region.name())],
     )?;
 
+    let s3_client = aws_util::s3::client(aws_info.clone()).await.err_hint(
+        "Creating S3 client",
```
Comment (benesch) [inline src/testdrive/src/action.rs:558]: testdrive starts error contexts in lowercase: "creating S3 client"

File: src/testdrive/src/action.rs [hunk 18]
```diff
@@ -472,11 +573,13 @@ pub async fn create_state(
         kafka_config,
         kafka_producer,
         kafka_topics,
-        aws_region: config.aws_region.clone(),
+        aws_region: aws_info.region,
         aws_account: config.aws_account.clone(),
-        aws_credentials: config.aws_credentials.clone(),
+        aws_credentials: aws_info.credentials.unwrap().into(),
```
Comment (benesch) [inline src/testdrive/src/action.rs:578]: What's up with this `unwrap`?

File: src/testdrive/src/action.rs [hunk 19]
```diff
@@ -57,6 +59,7 @@ pub struct Config {
 }
 
 pub struct State {
+    /// A random number to distinguish each TestDrive run
```
Comment (benesch) [inline src/testdrive/src/action.rs:62]: I'm not a big fan of adding arbitrary docstrings to things like this. This is private API; I don't really think it needs to be documented, to be honest.

File: test/testdrive/README.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
+TestDrive Tests
```
Comment (benesch) [inline test/testdrive/README.md:1]: It's "testdrive" (or "Testdrive" when English capitalization conventions demand it), but never "TestDrive"!

File: src/testdrive/src/main.rs [hunk 21]
```diff
@@ -82,38 +82,35 @@ async fn main() {
...
+                .aws_endpoint
+                // normalize endpoint, considering an empty string to be None
+                .and_then(|e| if e == "" { None } else { Some(e) })
+                .unwrap_or_else(|| "http://localhost:4566".into()),
         };
+        let account = "000000000000";
+        let credentials =
+            AwsCredentials::new("dummy-access-key-id", "dummy-secret-access-key", None, None);
+        (region, account.into(), credentials)
+    };
```
Comment (benesch) [inline src/testdrive/src/main.rs:113]: I'm confused about why this changed. What was wrong with the old logic?

File: test/testdrive/mzcompose.yml [hunk 22]
```diff
@@ -8,6 +8,58 @@
...
+    steps:
+      - step: start-services
+        services: [localstack]
+      - step: workflow
+        workflow: start-deps
+      - step: wait-for-tcp
+        host: localstack
+        port: 4566
+      - step: run
+        service: testdrive
```
Comment (benesch) [inline test/testdrive/mzcompose.yml:49]: Great idea!

File: test/testdrive/mzcompose.yml [hunk 23]
```diff
@@ -8,6 +8,58 @@
...
+    steps:
+      - step: start-services
+        services: [kafka, schema-registry, materialized]
+      - step: wait-for-tcp
+        host: schema-registry
+        port: 8081
+      - step: wait-for-tcp
+        host: kafka
+        port: 9092
+      - step: wait-for-mz
```
Comment (benesch) [inline test/testdrive/mzcompose.yml:61]: This seems like duplicate waiting. If we do this, can we get rid of the `wait-for-it` commands in the entrypoint, and also in the `testdrive` image?

File: test/testdrive/mzcompose.yml (hunk 1 above)
Comment (benesch) [reply test/testdrive/mzcompose.yml:20]: I definitely don't think `testdrive` the binary should be looking at an environment variable that exists just to work around limitations in mzcompose! I'm not sure I follow the `command_target` proposal either.

Can we instead just change the testdrive entrypoint to `cd` into `test/testdrive` first, and drop the `test/testdrive` prefix in the default command? Then everything will be automatically relative to `test/testdrive`.

**Review by benesch — COMMENTED** [review]
Whew, ok, first round of comments is in!

File: test/testdrive/README.md (hunk 20 above)
Comment (quodlibetor) [reply test/testdrive/README.md:1]: Good to know.

File: src/testdrive/src/main.rs (hunk 21 above)
Comment (quodlibetor) [reply src/testdrive/src/main.rs:113]: The `None` is overprecise: we actually want to use the endpoint specifically when the region is not valid, whether or not the endpoint is set.

In particular, that allows setting the endpoint in tests and leaving it present.

File: test/testdrive/mzcompose.yml (hunk 23 above)
Comment (quodlibetor) [reply test/testdrive/mzcompose.yml:61]: yes, we can and I'd prefer it, I just didn't want to bust anyone's workflow. That said, I can just make an announcement.

File: test/testdrive/mzcompose.yml (hunk 1 above)
Comment (quodlibetor) [reply test/testdrive/mzcompose.yml:20]: I'll try that, yeah.

Comment (quodlibetor) [reply test/testdrive/mzcompose.yml:20]: > I'm not sure I follow the command_target proposal either.

The idea is we could write:

```yaml
mzworkflows:
  local:
    command_target: testdrive
    steps:
      - step: run
        service: testdrive
```

and then if you run `mzcompose run ci --ci-output joins.td` the `--ci-output joins.td` will get
forwarded to `testdrive`. This is forward-compatible with a hypothetical `--command=testdrive='--ci-output joints.td'` thing that we could also do.

Comment (quodlibetor) [reply test/testdrive/mzcompose.yml:20]: `TD_TEST` with a directory change works, though, so I'll leave that hypothetical hunk of work off for a future person to think about.

File: src/testdrive/src/action.rs (hunk 12 above)
Comment (quodlibetor) [reply src/testdrive/src/action.rs:187]: Yeah a repeating error would do that, switched this whole thing to use `ore::retry`

File: src/testdrive/src/action.rs (hunk 18 above)
Comment (quodlibetor) [reply src/testdrive/src/action.rs:578]: Similar to above, changed it to an expect with explanation.

File: src/testdrive/src/error.rs (hunk 11 above)
Comment (quodlibetor) [reply src/testdrive/src/error.rs:51]: I support that! I can't think of a much better way to do this, either.

File: src/testdrive/src/error.rs (hunk 9 above)
Comment (quodlibetor) [reply src/testdrive/src/error.rs:119]: No it's not great, but the previous implementation just threw away information. _Usually_ I would expect only one error, or all errors to have the same root cause so it's hopefully not a huge detail in the usual case.

I could switch the separator, or wrap them in parens or square brackets instead, which is uglier but less likely to be obviously confusing.

File: src/testdrive/src/error.rs (hunk 10 above)
Comment (quodlibetor) [reply src/testdrive/src/error.rs:208]: weird.

File: src/testdrive/src/action.rs (hunk 19 above)
Comment (benesch) [reply src/testdrive/src/action.rs:62]: Yeah, I'm sorry if this feels really nitpicky, but what I specifically find undesirable is having structs that are sporadically documented like this.  A struct with just a smattering of its fields documented just looks unfinished and wrong to my eye, and I find it very hard to see past that when I'm reading the code. For me, adding a one-off comment like this takes this struct from a polished and succinct representation of testdrive's state, though perhaps a somewhat underdocumented one, to one that is looks WIP, waiting for someone to come along and describe what the rest of the fields mean.

I just landed a PR I think you'll like that cleans up the documentation for testdrive's public API (https://github.com/MaterializeInc/materialize/pull/5308), so that mtrlz.dev/api/rust will render nicely. It adds a a complete set of comments to the `Config` struct.

To back up for a second, I'm unconvinced of the value of this sort of documentation for internal API. I think the idea that folks can pop into a codebase and have a complete set of accurate docstrings to show them around is unrealistic for any real-world code, and that the process of building a mental model of code is much more about spending time internalizing the various objects and how they fit together. In my experience, making things work more obviously is worth 100x more writing down how and why they behave non-obviously.

Anyway, I'm willing to put that opinion on hold, because I realize it is not widely popular with many other folks here. But my ask in return is that we are very intentional and precise when we do add documentation.

Specifically, if we're going to document `State`, here is how I would prefer for it to look!

```rust
/// The state associated with executing a single testdrive script.
///
/// A state is created at the beginning of each testdrive script execution,
/// then inspected and possibly mutated by each action in turn. The various
/// `reset` methods can be invoked at various points in the state's lifecycle
/// to clean up any external state modified by the script's actions.
pub struct State {
    /// A random number that uniquely identifies this state.
    ///
    /// Its value is exposed to scripts via the `testdrive.seed` variable. The
    /// seed is intended to be incorporated in the naming of objects stored in
    /// external state, so that multiple testdrive scripts can target the same
    /// external state simultaneously without conflicting. For example,
    /// `kafka::CreateTopicAction` incorporates the seed into the topic name.
    seed: u32,
    /// A temporary directory into which the script can emit data.
    ///
    /// The directory is destroyed when the state is dropped. The path of the
    /// directory is exposed to scripts via the `testdrive.temp_dir` variable.
    temp_dir: tempfile::TempDir,
    /// The optional path to the catalog backing the materialized instance to
    /// which testdrive connects.
    ///
    /// If present, actions can inspect the catalog contents as they like, to
    /// validate that the catalog matches their expectations.
    materialized_catalog_path: Option<PathBuf>,
    /// The address of the materialized instance to connect to.
    materialized_addr: String,
    /// A PostgreSQL client that is connected to the materialized instance
    /// specified by `materialized_addr`.
    pgclient: tokio_postgres::Client,
    /// The URL of a Confluent Schema Registry.
    ///
    /// The URL is exposed to scripts via the `testdrive.schema-registry-url`
    /// variable.
    schema_registry_url: Url,
    /// A schema registry client that is connected to the schema registry
    /// specified by `schema_registry_url`.
    ccsr_client: ccsr::Client,
    /// The address of a Kafka broker to connect to.
    ///
    /// The address is exposed to scripts via the `testdrive.kafka-addr`
    /// variable.
    kafka_addr: String,
    /// A Kafka admin client connected to the Kafka broker specified by `kafka_addr`.
    kafka_admin: rdkafka::admin::AdminClient<rdkafka::client::DefaultClientContext>,
    /// The options that should be used when making requests of the admin
    /// broker named in `kafka_admin`.
    kafka_admin_opts: rdkafka::admin::AdminOptions,
    /// The configuration options that should be used when creating Kafka
    /// clients that target the Kafka broker specified by `kafka_addr`.
    kafka_config: ClientConfig,
    /// A Kafka producer connected to the Kafka broker specified by
    /// `kafka_addr`.
    kafka_producer: rdkafka::producer::FutureProducer<rdkafka::client::DefaultClientContext>,
    /// A map from name to partition count for all Kafka topics created with
    /// this state.
    kafka_topics: HashMap<String, usize>,
    /// The AWS region to connect to.
    ///
    /// The name of the AWS region and the custom endpoint, if any, are exposed
    /// to the testdrive script as `testdrive.aws-region` and
    /// `testdrive.aws-endpoint`, respectively.
    aws_region: rusoto_core::Region,
    /// The account number for the account associated with AWS credentials
    /// specified in `aws_credentials`.
    aws_account: String,
    /// The credentials to use when connecting to AWS services.
    ///
    /// The access key ID, secret access key, and token within are exposed to
    /// scripts via the `testdrive.aws-access-key-id`,
    /// `testdrive.aws-secret-access-key`, and `testdrive.aws-token` variables,
    /// respectively.
    aws_credentials: AwsCredentials,
    /// A Kinesis client authenticated using `aws_credentials`.
    kinesis_client: KinesisClient,
    /// The names of the Kinesis streams created with this state.
    ///
    /// Calling [`State::reset_kinesis`] will attempt to delete these Kinesis
    /// streams.
    kinesis_stream_names: Vec<String>,
}
```

File: test/testdrive/mzcompose.yml (hunk 23 above)
Comment (benesch) [reply test/testdrive/mzcompose.yml:59]: Oh, you know, the other thing we could do is rename `local` to `testdrive`, and rename the `testdrive` service to `testdrive-container` or something like that.

File: src/testdrive/src/main.rs (hunk 21 above)
Comment (benesch) [reply src/testdrive/src/main.rs:120]: So the idea with the old logic was that if you didn't specify an endpoint, then we'd see if the region was valid. If _either_ the region was invalid, or you specified an endpoint, then we'd take your `(region, endpoint)` pair literally.

The new code looks like it'll silently ignore the endpoint if you specify a valid AWS region, which doesn't seem right?

File: src/testdrive/src/error.rs (hunk 9 above)
Comment (benesch) [reply src/testdrive/src/error.rs:122]: Yeah, gotcha, it's probably fine. 🤷🏽 We can improve it if it becomes an issue!

File: src/testdrive/src/action.rs (hunk 19 above)
Comment (benesch) [reply src/testdrive/src/action.rs:63]: Brandon and I talked offline; I'm going to periodically swoop in and use the comments Brandon adds to determine where things are underdocumented and add more cohesive docs. Therefore this is fine for now!

**Comment by benesch** [comment]
Can I help here? It might be useful to merge the first two commits to get things working locally, and then we can iterate on the CI stuff afterwards.

**Comment by quodlibetor** [comment]
> Can I help here?

There are errors in Kinesis-auth-in-CI, I think I'm narrowing down the possible causes, I did futz with a lot of the auth code to centralize it, but Kinesis works slightly differntly to other servics.

File: src/sql/src/plan/statement/ddl.rs [hunk 24]
```diff
@@ -404,38 +403,18 @@ pub fn plan_create_source(
...
-            };
+            let region = arn
+                .region
+                .ok_or_else(|| anyhow!("Provided ARN does not include an AWS region"))?;
 
+            let aws_info = aws_connect_info(&mut with_options, Some(region))?;
+            futures::executor::block_on(aws_util::aws::validate_credentials(
+                aws_info.clone(),
+                Duration::from_secs(1),
+            ))?;
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:414]: Eek, blocking in the executor like this is not ok! It'll prevent any other SQL query from getting processed for 1s.

**Comment by benesch** [comment]
There's a lot of good stuff in here, but also some stuff that I think deserves closer scrutiny. It's hard for me to tell what's improving the current state of `main` and what's fixing issues introduced within this PR. Can we start with splitting out your original version of the test into its own PR (https://github.com/MaterializeInc/materialize/pull/5331/commits/b101f57619d92005ac09b968246610ddd051087c) and try to get that merged first, without the additional localstack/credential verification changes?

**Comment by benesch** [comment]
> There's a lot of good stuff in here, but also some stuff that I think deserves closer scrutiny. It's hard for me to tell what's improving the current state of `main` and what's fixing issues introduced within this PR. Can we start with splitting out your original version of the test into its own PR ([b101f57](https://github.com/MaterializeInc/materialize/commit/b101f57619d92005ac09b968246610ddd051087c)) and try to get that merged first, without the additional localstack/credential verification changes?

I took a stab at this in #5460.

**Comment by quodlibetor** [comment]
> Can we start with splitting out your original version of the test into its own PR

I see that you've already done this, but yeah I would be happy to as soon as I can get the tests to pass, I'm not planning on looking at that until I know if I need to change something in one of the earlier commits to git it to pass.

We should hold off on reviewing this (or my future PR) until I've got the tests passing.

**Comment by quodlibetor** [comment]
this has been broken into several other PRs.
